Automattic/jetpack
# JSON API: Add custom taxonomies support

This PR adds support for custom taxonomies to the JSON API.  This is a port of @aduth's equivalent changes from WP.com and is required for Jetpack support for the upcoming custom post types features in Calypso.

New API endpoints:

- `GET /rest/v1.1/sites/$site/post-types/$type/taxonomies` - list taxonomies associated with a post type
- `GET /rest/v1.1/sites/$site/taxonomies/$taxonomy/terms` - list terms in a taxonomy
- `GET /rest/v1.1/sites/$site/taxonomies/$taxonomy/terms/slug:$slug` - get a term
- `POST /rest/v1.1/sites/$site/taxonomies/$taxonomy/terms/new` - create a new term
- `POST /rest/v1.1/sites/$site/taxonomies/$taxonomy/terms/slug:$slug` - edit a term
- `POST /rest/v1.1/sites/$site/taxonomies/$taxonomy/terms/slug:$slug/delete` - delete a term

Also, the endpoints to create and update a post support a new `terms` property which is a map of taxonomies to arrays of term names.

The new API features have unit tests on WP.com, but this is a massive PR, so I've created a test script to  run though the new capabilities on a Jetpack site.  Here's the script:  https://gist.github.com/nylen/246da68b6f309ff8392e0321e462bf12

Here's how to use it to test this PR:

- Obtain an OAuth2 token that is valid for your Jetpack site.  See the [documentation](https://developer.wordpress.com/docs/oauth2/) or use [this tool](http://nylen.io/get-wpcom-token/) to help.
- Install and activate [this simple plugin](https://cloudup.com/cIsFtBaQaGA) on your Jetpack site.  It registers a `book` custom post type with a `genre` taxonomy.
- Run the test script, passing it your site ID and token secret on the command line.

More info: p3hLNG-G9-p2
